### DISCUSSION TITLE
Using "images" folders to use existing resources and create customized images

### DISCUSSION THREAD

**Discussion by sharadbharadia** [actions/virtual-environments#3578] [Image configuration]
We already have Azure DevOps project setup which customizes images and pushes them to shared image galley.  We are trying to use "images\linux" folder from https://github.com/actions/virtual-environments repo as is to create customized `Ubuntu2004' image. However our pipeline fails while executing below code

```
{
        "type": "shell",
        "scripts": [
            "{{template_dir}}/scripts/installers/complete-snap-setup.sh"
        ],
        "environment_vars": [
            "HELPER_SCRIPTS={{user `helper_script_folder`}}"
        ],
        "execute_command": "sudo sh -c '{{ .Vars }} {{ .Path }}'"
},
```

Error message says:

```
1623504910,,ui,say,==> azure-arm: Provisioning with shell script: D:\a\1\s\VMs\linux/scripts/installers/configure-environment.sh
1623504910,,ui,message,    azure-arm: ImageVersion=dev
1623504910,,ui,message,    azure-arm: ImageOS=ubuntu20
1623504910,,ui,message,    azure-arm: ACCEPT_EULA=Y
1623504910,,ui,message,    azure-arm: XDG_CONFIG_HOME=$HOME/.config
1623504910,,ui,message,    azure-arm: AGENT_TOOLSDIRECTORY=/opt/hostedtoolcache
1623504910,,ui,message,    azure-arm: vm.max_map_count=26214
1623504910,,ui,say,==> azure-arm: Provisioning with shell script: D:\a\1\s\VMs\linux/scripts/installers/complete-snap-setup.sh
1623504911,,ui,error,==> azure-arm: /imagegeneration/helpers/etc-environment.sh: line 6: $'\r': command not found
```

Spent quite time to diagnose the issue but couldn't. Any ideas why it is failing ?

ADO pipeline is hosted on `Microsoft Hosted Agent : vmImage: 'windows-2019'` which calls a PowerShell script `Pack-Image.ps1` to create the image.

**Comment by sharadbharadia** [comment]
Ok. I solved the issue. I had to install `dos2unix`  
```
"provisioners": [
        {
          "type": "shell",
          "inline": [
          "sudo apt update -y",
          "sudo apt full-upgrade -y",
          "sudo apt clean all",
          "sudo apt-get install dos2unix"
          ]
      },
```

and then added below code to packerfile 
```
{
          "type": "shell",
          "inline": [
              "chmod +x -R {{user `image_folder`}}/*",
              "dos2unix $(find {{user `image_folder`}} -type f)"
          ],
          "execute_command": "sudo sh -c '{{ .Vars }} {{ .Path }}'"
},
```

**Reply by maxim-lobanov** [reply]
Thank you for report.
The issue looks pretty strange. Looks like files have incorrect line-endings (`CRLF` (Win) vs `LF` (Unix)) in your local instance of `virtual-environments` repository.

I have checked and all files inside Linux folder has correct line-endings in our repo. Also, `.gitattributes` is configured properly: https://github.com/actions/virtual-environments/blob/main/.gitattributes#L1
